src/java/org/apache/cassandra/io/sstable/StorageHandler.java

```diff
@@ -0,0 +1,149 @@
+ * work, such as selecting only part of the remote sstables available, or adding new ones when offline compaction
+ * has run. This behaviour can be implemented in a sub-class that can be set with {@link #remoteStorageHandler}.
+ * <p/>
+ * Local sytem tables will always use the default storage handler, {@link DefaultStorageHandler}
+ * whereas user tables will use {@link #remoteStorageHandler} if one is set, or the default handler if none
+ * has been set.
+ * <p/>
+ */
+public abstract class StorageHandler
+{
+    private final static String remoteStorageHandler = REMOTE_STORAGE_HANDLER.getString();
+
+    public enum ReloadReason
+    {
+        /** New nodes joined or left */
+        TOPOLOGY_CHANGED,
+        /** Data was truncated */
+        TRUNCATION,
+        /** Ordinary periodic reload */
+        PERIODIC,
+        /** Data was replayed either from the commit log or a batch log */
+        DATA_REPLAYED,
+        /** When repair task started */
+        REPAIR,
+        /** A request over forced by users to reload. */
+        USER_REQUESTED,
+    }
+
+    protected final TableMetadataRef metadata;
+    protected final Directories directories;
+    protected final Tracker dataTracker;
+
+    public StorageHandler(TableMetadataRef metadata, Directories directories, Tracker dataTracker)
+    {
+        Preconditions.checkNotNull(directories, "Directories should not be null");
+
+        this.metadata = metadata;
+        this.directories = directories;
+        this.dataTracker = dataTracker;
+    }
+
+    /**
+     * @return true if the node is ready to serve data for this table. This means that the
+     *         node is not bootstrapping and that no data may be missing, e.g. if sstables are
+     *         being downloaded from remote storage or streamed from other nodes then isReady()
+     *         would return false. Generally, user read queries should not succeed if this method
+     *         returns false.
+     */
+    public abstract boolean isReady();
+
+    /**
+     * Load the initial sstables into the tracker that was passed in to the constructor.
+     *
+     * @return the sstables that were loaded
+     */
+    public abstract Collection<SSTableReader> loadInitialSSTables();
+
+    /**
+     * Reload any sstables that may have been created and not yet loaded. This is normally
+     * a no-op for the default local storage, but for remote storage implementations it
+     * signals that sstables need to be refreshed.
+     */
+    public abstract void reloadSSTables(ReloadReason reason);
+
+    /**
+     * This method determines if the {@link ColumnFamilyStore} should perform background compactions,
+     * the default behavior for local sstables.
+     * <p/>
+     * If sstables are remote, compaction can be done offline, in which case this method should return false.
+     * <p/>
+     * @return true if auto-compaction should be enabled, false otherwise
+     */
+    public abstract boolean enableAutoCompaction();
```

Comment:
> So maybe this should just be called isRemote as it simply returns true for remote and false for local storage handlers?

that means `Stargazer` needs to know about remote storage and knows that remote storage doesn't support auto-compaction

Reply:
So the meaning is like `isAutoCompactionSupported`? I thought that FS does not have anything to do with auto-compaction - how it can be related?

Reply:
it means `isAutoCompactionEnabled()`.  

> I thought that FS does not have anything to do with auto-compaction - how it can be related?

I mean the decision should not be made on stargazer side, like you mentioned `called isRemote as it simply returns true for remote and false for local storage handlers?`.

The compaction logic for remote files are moved to `CNDB compactor`, thus auto-compaction is disabled for table using remote storage..

Reply:
This is too confusing to me - we have two sources of truth now - `strategyContainer` and `storageHandler` - both can return different results - what is the relation between them?

Reply:
How about renaming it to `supportsAutoCompactions()`? auto-compaction should be run if the storage handler supports and it's enabled in comapction param in compaction container

Reply:
it is better, but still it does not feel like it is a property of storage handler - you actually said the same if I got you right, the compaction logic in compactor, which makes a decision based on the fact whether the storage is local or remote - `isRemote` is pure storage property and the compactor could do whatever it wants based on that, for example, why not add a boolean param to `CompactionStrategyContainer.reload` and pass `storageHandler.isRemote` here:

```
        ...
        this.strategyContainer = strategyFactory.reload(null,
                                                        metadata.get().params.compaction,
                                                        CompactionStrategyContainer.ReloadReason.FULL,
                                                        this.storageHandler.isRemote());
        ...

        if (DISABLED_AUTO_COMPACTION_PROPERTY.getBoolean())
        {
            ...
            this.strategyContainer.disable();
        }
```

Reply:
updated to `isRemote` and pass to `strategyContainer`

Reply:
thank you